src/cir.rs

```diff
@@ -0,0 +1,365 @@
+    }
+}
+
+pub trait Translate {
+    type Output;
+    fn translate(&self) -> Self::Output;
+}
+
+impl Translate for Expression {
+    type Output = CirProgram;
+    fn translate(&self) -> Self::Output {
+        let mut cir = CirProgram::new();
+        cir_translate_helper(self, &mut cir);
+        cir.instructions.shrink_to_fit(); // shrink the memory
+        cir
+    }
+}
+
+fn cir_translate_helper(exp: &Expression, cir: &mut CirProgram) -> usize {
+    match exp {
+        Expression::Logical(logic_exp) => match logic_exp.as_ref() {
+            LogicalExpression::And(l, r) => {
+                let left = match l {
+                    Expression::Logical(_logic_exp) => {
+                        CirOperand::Index(cir_translate_helper(l, cir) - 1)
+                    }
+                    Expression::Predicate(p) => CirOperand::Predicate(p.clone()),
+                };
+
+                let right = match r {
+                    Expression::Logical(_logic_exp) => {
+                        CirOperand::Index(cir_translate_helper(r, cir) - 1)
+                    }
+                    Expression::Predicate(p) => CirOperand::Predicate(p.clone()),
+                };
+                let and_ins = AndIns { left, right };
+                cir.instructions.push(CirInstruction::AndIns(and_ins));
+            }
+            LogicalExpression::Or(l, r) => {
+                let left = match l {
+                    Expression::Logical(_logic_exp) => {
+                        CirOperand::Index(cir_translate_helper(l, cir) - 1)
+                    }
+                    Expression::Predicate(p) => CirOperand::Predicate(p.clone()),
+                };
+
+                let right = match r {
+                    Expression::Logical(_logic_exp) => {
+                        CirOperand::Index(cir_translate_helper(r, cir) - 1)
+                    }
+                    Expression::Predicate(p) => CirOperand::Predicate(p.clone()),
+                };
+                let or_ins = OrIns { left, right };
+                cir.instructions.push(CirInstruction::OrIns(or_ins));
+            }
+            LogicalExpression::Not(r) => {
+                let right: CirOperand = match r {
+                    Expression::Logical(_logic_exp) => {
+                        CirOperand::Index(cir_translate_helper(r, cir) - 1)
+                    }
+                    Expression::Predicate(p) => CirOperand::Predicate(p.clone()),
+                };
+                let not_ins = NotIns { right };
+                cir.instructions.push(CirInstruction::NotIns(not_ins));
+            }
+        },
+        Expression::Predicate(p) => {
+            cir.instructions.push(CirInstruction::Predicate(p.clone()));
+        }
+    }
+    cir.instructions.len()
+}
+
+fn execute_helper(
+    cir_instructions: &[CirInstruction],
+    index: usize,
+    ctx: &mut Context,
+    m: &mut Match,
+) -> bool {
+    match &cir_instructions[index] {
```

Comment:
I recommend reordering the instructions. Reading the array in inverse order hurts the CPU's caching performance.